modules/renter/contractor/contractor.go

```diff
@@ -81,41 +76,76 @@ func (c *Contractor) SetAllowance(a modules.Allowance) error {
 		return errors.New("renew window must be less than period")
 	}
 
-	err := c.formContracts(a)
+	// check that allowance is sufficient to store at least one sector
+	numSectors, err := maxSectors(a, c.hdb)
 	if err != nil {
 		return err
+	} else if numSectors == 0 {
+		return errInsufficientAllowance
 	}
 
-	// Set the allowance.
+	// prevent ProcessConsensusChange from renewing contracts until we have
+	// finished
+	c.contractLock.Lock()
+	defer c.contractLock.Unlock()
+
 	c.mu.Lock()
-	c.allowance = a
-	err = c.saveSync()
+	endHeight := c.blockHeight + a.Period
+
+	// check if allowance has different period or funds; if so, we should
+	// renew existing contracts with the new allowance.
+	// TODO: compare numSectors instead of allowance.Funds?
+	var renewSet []modules.RenterContract
+	if a.Period != c.allowance.Period || a.Funds.Cmp(c.allowance.Funds) != 0 {
+		for _, contract := range c.contracts {
+			renewSet = append(renewSet, contract)
+		}
+		// TODO: dangerous -- figure out a safer approach later
+		c.contracts = map[types.FileContractID]modules.RenterContract{}
```

Comment:
In the new approach, as soon as you set a different allowance, all the old contracts are "deleted" (moved from `c.contracts` to `renewSet`). This gives us a clean slate to work with, so that we know all contracts in `c.contracts` were formed with the same allowance.
The downside is that if you shutdown before `SetAllowance` returns, you might have deleted a contract without renewing it. So when you start up again, you'll need to form a new contract instead of renewing.

Another unresolved issue is that contracts can still be "heterogeneous" with respect to their end height. That is, if you call `SetAllowance` once and it forms 3 out of 6 contracts, and then call it again 10 blocks later, you'll wind up with 3 contracts ending at one height and 3 ending at another. I don't think this is a huge issue, but it's worth fixing for the sake of robustness.